### PR TITLE
fix(python): fix stacklevel of some deprecation warnings

### LINKED ISSUE #8088: DataFrame.pivot - incorrect DeprecationWarning stack level
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

`DeprecationWarning` shows incorrect location,

### Reproducible example

```python
import polars as pl

def test_func():
    df = pl.DataFrame(
        {
            "foo": ["one", "one", "one", "two", "two", "two"],
            "bar": ["A", "B", "C", "A", "B", "C"],
            "baz": [1, 2, 3, 4, 5, 6],
        }
    )
    df.pivot(values="baz", index="foo", columns="bar")

def main():
    test_func()

if __name__ == "__main__":
    main()  # <<< DeprecationWarning reported here
```

Reports warning location:

```
test.py:17: DeprecationWarning: In a future version of polars, the default `aggregate_function` will change from `'first'` to `None`. Please pass `'first'` to keep the current behaviour, or `None` to accept the new one.
  main()
```

### Expected behavior

Correct warning location.

### Installed versions

<details>

```
---Version info---
Polars: 0.17.0
Index type: UInt32
Platform: Windows-10-10.0.19045-SP0
Python: 3.10.8 (tags/v3.10.8:aaaf517, Oct 11 2022, 16:50:30) [MSC v.1933 64 bit (AMD64)]
---Optional dependencies---
numpy: 1.23.5
pandas: 1.5.3
pyarrow: 11.0.0
connectorx: <not installed>
deltalake: <not installed>
fsspec: 2023.1.0
matplotlib: 3.7.1
xlsx2csv: <not installed>
xlsxwriter: <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #8088 

demo:
```
# tmp.py
import polars as pl
df = pl.DataFrame(
    {
        "foo": ["one", "one", "one", "two", "two", "two"],
        "bar": ["A", "B", "C", "A", "B", "C"],
        "baz": [1, 2, 3, 4, 5, 6],
    }
)
df.pivot(values="baz", index="foo", columns="bar")
```

```console
$ python tmp.py 
/home/marcogorelli/polars-dev/py-polars/tmp.py:9: DeprecationWarning: In a future version of polars, the default `aggregate_function` will change from `'first'` to `None`. Please pass `'first'` to keep the current behaviour, or `None` to accept the new one.
  df.pivot(values="baz", index="foo", columns="bar")
```

### CODE REVIEW

**Comment by ritchie46** [comment]
How should the stacklevels be determined? Just trying it out?

**Comment by MarcoGorelli** [comment]
either that, or with something like

https://github.com/pandas-dev/pandas/blob/ab89c53f48df67709a533b6a95ce3d911871a0a8/pandas/util/_exceptions.py#L30-L51

. there would be a slight perf hit compared with computing the number manually and just hardcoding it, but it would just be on the Python side, shouldn't be in any loop, and only for deprecated behaviour anyway. OK to pull it in?

**Comment by 2-5** [comment]
One more automatic way is to manually walk up the stack until you exit the `polars` package.

**Comment by MarcoGorelli** [comment]
I think this might be warranted, actually

If I have
```python
pl.DataFrame({'ts': pl.Series(['2020-01-01'])}).select(pl.col('ts').str.strptime(pl.Datetime, tz_aware=False))
pl.Series(['2020-01-01']).str.strptime(pl.Datetime, tz_aware=False)
```

then only the first line shows the warning, because the warning requires different stacklevels in the two cases

**Comment by MarcoGorelli** [comment]
I timed `find_stacklevel`, and it seems really insignificant:
```
ncalls 1
tottime 8.02e-06
percall 8.02e-06
cumtime 3.474e-05
percall 3.474e-05
filename:lineno(function) various.py:343(find_stacklevel)
```

I got this by running:
- python -m cProfile  -o out.prof tmp.py
- snakeviz out.prof --server
and then looking for `find_stacklevel` in the output

**Comment by stinodego** [comment]
I'd be very much in favor of using `find_stacklevel`. We shouldn't have developers guessing at the correct stack-level to use.

And if we have that util, we can get rid of the `check_stacklevels` script, right?

**Comment by MarcoGorelli** [comment]
Sure, done!

I simplified the script to just check that `stacklevel=` isn't called with a constant, pointing developers towards `find_stacklevel` instead

**Review by stinodego — APPROVED** [review]
This is a nice improvement and should be merged!

I only wonder how this interacts with the `deprecated_alias` and `deprecate_nonkeyword_arguments` utils. Those can still go wrong when not explicitly specifying the stack-level, right?

**Comment by 2-5** [comment]
It might be worth considering moving the raising of `DeprecationWarning` inside it's own helper function:


```python
warnings.warn(
    f"`{type(obj).__name__}.{item}` has been renamed; this"
    f" redirect is temporary, please use `.{new_item}` instead",
    category=DeprecationWarning,
    stacklevel=find_stacklevel(),
)
```

becomes

```python
warn_deprecation(
    f"`{type(obj).__name__}.{item}` has been renamed; this"
    f" redirect is temporary, please use `.{new_item}` instead",
)
```

**Comment by MarcoGorelli** [comment]
> I only wonder how this interacts with the deprecated_alias and deprecate_nonkeyword_arguments utils. Those can still go wrong when not explicitly specifying the stack-level, right?

thanks - I've put `find_stacklevel` inside them, and removed `stacklevel` as an argument, so it'll be correctly set (automatically)

> It might be worth considering moving the raising of DeprecationWarning inside it's own helper function:

maybe, although `stacklevel` also needs setting for other kinds of warnings

**Review by stinodego — APPROVED** [review]

File: py-polars/scripts/check_stacklevels.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+# even on Windows, so it's OK to list them like that.
+EXCLUDE = frozenset(["polars/utils/polars_version.py"])
+
+
+class StackLevelChecker(NodeVisitor):
+    def __init__(self, file) -> None:
+        self.file = file
+        self.violations = set()
+
+    def visit_Call(self, node: ast.Call) -> None:
```
Comment (ritchie46) [inline py-polars/scripts/check_stacklevels.py:22]: I think `c` should be lowercase?

Comment (MarcoGorelli) [reply py-polars/scripts/check_stacklevels.py:22]: the class is `ast.Call`, and the format to override visitors is `visit_classname` https://docs.python.org/3/library/ast.html#ast.NodeVisitor.visit

the ast module is a bit odd in this respect, it's part of stdlib but it breaks pep8..

Comment (ritchie46) [reply py-polars/scripts/check_stacklevels.py:22]: Right, maybe add a comment about that? I am sure I will wonder about this again the second time I read this. ^^

Comment (ritchie46) [reply py-polars/scripts/check_stacklevels.py:22]: A nevermind. I see the signature with `ast.Call`. :+1:

**Comment by alexander-beedie** [comment]
Nice one 👍

**Pull request merged**